### ISSUE TITLE
eleventy.before callback runs after ignores are processed

### ISSUE THREAD

**Issue by TheDocTrier** [11ty/eleventy#2207]
**Describe the bug**

If a callback is added to the `eleventy.before` event, that callback is executed after Eleventy has already processed the `eleventyConfig.ignores` set.

**To Reproduce**

Create a template `ignore.md`.
Add callback on `eleventy.before` event which adds template's path to `ignores`.
Build the site.
The template will be rendered.

```js
module.exports = function (eleventyConfig) {
  eleventyConfig.on("eleventy.before", function () {
    eleventyConfig.ignores.add("ignore.md");
  });
};
```

If `eleventy --serve` is used, then the template will be ignored on subsequent rebuilds.

Here is a simple repo that reproduces the behavior: https://github.com/TheDocTrier/eleventy-before-build-ignore 

**Expected behavior**

The `eleventy.before` callbacks are run before Eleventy processes the ignores set such that modifications to it are reflected in the build.

**Environment:**

- OS and Version: Ubuntu 18.04
- Eleventy Version 8.1.2

**Edit**
Update issue and example to use `eleventy.before` instead of the deprecated `beforeBuild`.

**Comment by pdehaan** [comment]
Quick note since I was looking at the docs for this earlier today, but it appears that the `.on("beforeBuild", ...)` is deprecated, per https://www.11ty.dev/docs/events/#eleventy.before, and you might want to retry w/ `.on("eleventy.before", ...)`:

> ### `eleventy.before` <i>(New in v1.0.0)</i>
> - Previously known as the now deprecated (but not removed) `beforeBuild` <i>(New in v0.11.1)</i> event.


```diff
-eleventyConfig.on("beforeBuild", function () {
+eleventyConfig.on('eleventy.before', async () => {
```

**Comment by TheDocTrier** [comment]
Thank you @pdehaan for pointing that out. The problem is persistent with the new label and I have updated the example repository to reflect this.

**Comment by pdehaan** [comment]
Yep, can confirm on v1.0.0 that calling `eleventyConfig.ignores.add()` from within `.on("eleventy.before")` doesn't work, but works if you add it in the top-level eleventy config.

```js
  eleventyConfig.ignores.add("ignore.md"); // WORKS

  eleventyConfig.on("eleventy.before", function () {
    eleventyConfig.ignores.add("ignore.md"); // DOESNT WORK
  });
```

**Comment by TheDocTrier** [comment]
Yeah, I've used that as a work-around in my Eleventy ignore plugin:
https://github.com/TheDocTrier/eleventy-plugin-ignore/blob/561dc0d3fa161fccf3f86f5980436d93e60e2d67/src/eleventy.ts#L48-L49
```js
updateIgnores();
eleventyConfig.on("eleventy.before", updateIgnores);
```

**Comment by zachleat** [comment]
Hmmmmmm—I think there is some confusion here and maybe I should add some additional clarification to the docs, perhaps on both https://www.11ty.dev/docs/events/ and https://www.11ty.dev/docs/advanced-order/

Configuration is the very first thing that happens, before anything else. It is needed in various constructors in the code (for good or bad, see #2103).

So, to be clear the `eleventy.before` is called _after_ configuration is initialized. However, especially in plugin land, you probably shouldn’t use this event at all?

The plugin API will run your plugin callback function every build, see this reduced test case:

![image](https://user-images.githubusercontent.com/39355/154327539-b3261014-e5c9-479d-a3d9-c59f0d307fe9.png)

Is there some requirement I’m missing here?

**Comment by zachleat** [comment]
Sorry, I misspoke a bit there (stand by 😅)

**Comment by zachleat** [comment]
`eleventy.beforeConfig` is available in `v3.0.0-alpha.3` and will ship with v3.0.

It was originally added as an async mechanism to modify configuration in synchronous config callbacks.

**Comment by zachleat** [comment]
Temporary preview docs deploying: https://11ty-website-git-v3-11ty.vercel.app/docs/events/#eleventy.beforeconfig-pre-release-only-v3.0.0-alpha.3
